File: gke_customization/gke_survey/page/questionnaire_runner/questionnaire_runner.js
let page;
let template_field;

frappe.pages["questionnaire-runner"].on_page_load = function(wrapper) {

    page = frappe.ui.make_app_page({
        parent: wrapper,
        title: "Questionnaire Runner",
        single_column: true
    });

    template_field = page.add_field({
        label: "Questionnaire Template",
        fieldname: "questionnaire",
        fieldtype: "Link",
        options: "Questionnaire Template"
    });

    page.set_primary_action("Load Questionnaire", load_questionnaire);

    $(page.body).append(`
        <div id="questionnaire-container" class="mt-4"></div>
    `);
};

function load_questionnaire() {

    let questionnaire = template_field.get_value();

    if (!questionnaire) {
        frappe.msgprint("Please select Questionnaire");
        return;
    }

    frappe.call({
        method: "gke_customization.gke_survey.page.questionnaire_runner.questionnaire_runner.get_questionnaire",
        args: {
            template: questionnaire
        },
        callback: function(r) {

            render_questionnaire(r.message);

            // Change button after loading
            page.set_primary_action("Submit", submit_questionnaire);

        }
    });

}

function submit_questionnaire() {

    let answers = [];

    Object.values(questionnaire_controls).forEach(item => {

        answers.push({
            question: item.question.name,
            answer: item.control.get_value()
        });

    });

    frappe.call({
        method: "gke_customization.gke_survey.page.questionnaire_runner.questionnaire_runner.save_response",
        args: {
            questionnaire: template_field.get_value(),
            answers: answers
        },
        callback: function(r) {

            frappe.msgprint("Questionnaire Saved Successfully");

            console.log(r.message);

        }
    });

}


function render_questionnaire(data) {

    $("#questionnaire-container").empty();

    // Store controls globally for save later
    window.questionnaire_controls = {};

    data.sections.forEach(section => {

        let section_wrapper = $(`
            <div class="questionnaire-section mb-5">
                <h4>${section.section_name}</h4>
                <hr>
                <div class="row questions-row"></div>
            </div>
        `);

        $("#questionnaire-container").append(section_wrapper);

        let questions = data.questions.filter(q => {
            return q.section === section.name;
        });

        questions.forEach(question => {

            let col_size = question.width || 4; // default 3 columns

            let field_wrapper = $(`
                <div class="col-md-${col_size} mb-3">
                </div>
            `);

            section_wrapper.find(".questions-row").append(field_wrapper);

            let df = {
                label: question.question,
                fieldname: question.fieldname || frappe.scrub(question.question),
                fieldtype: question.field_type || "Data",
                reqd: question.mandatory || 0
            };

            // Select field
            if (question.field_type === "Select") {
                df.options = question.options;
            }

            // Link field
            if (question.field_type === "Link") {
                df.options = question.options;
            }

            // Check field
            if (question.field_type === "Check") {
                df.default = 0;
            }

            let control = frappe.ui.form.make_control({
                parent: field_wrapper,
                df: df,
                render_input: true
            });

            // Save reference for submit later
            window.questionnaire_controls[df.fieldname] = {
                control: control,
                question: question
            };
        });

    });
}

File: gke_customization/gke_survey/page/questionnaire_runner/questionnaire_runner.py
import frappe
import json


# /home/frappe/frappe-bench/apps/gke_customization/gke_customization/gke_survey/page/questionnaire_runner/questionnaire_runner.py.get_questionnaire
@frappe.whitelist(allow_guest=True)
def get_questionnaire(template):
    template_doc = frappe.get_doc("Questionnaire Template", template)

    sections = frappe.get_all(
        "Questionnaire Section",
        filters={
            "questionnaire": template
        },
        fields=[
            "name",
            "section_name",
            "sequence",
            "tab",
            "depends_on",
            "sub_section"
        ],
        order_by="sequence asc"
    )

    questions = frappe.get_all(
        "Questionnaire Question",
        filters={
            "questionnaire": template
        },
        fields=[
            "name",
            "question",
            "section",
            "field_type",
            "options",
            "required",
            "sequence","require_rating","require_remark"
        ],
        order_by="section asc, sequence asc"
    )

    return {
        "template": template_doc,
        "sections": sections,
        "questions": questions
    }

@frappe.whitelist(allow_guest=True)
def save_response(questionnaire,answers,branch=None,
    auditor_name=None,
    audit_month=None,
    audit_dates=None,
    unit_audited=None,
    gstin_of_unit=None,
    nature_of_operations=None,
    persons_met_during_audit=None,
    previous_audit_pending_points=None,
    serious_irregularities_found=None,
    fraud_suspected=None,
    matter_requiring_immediate_attention=None,
    employee_id=None,
    employee_name=None,
    department=None):

    if isinstance(answers, str):
        answers = json.loads(answers)

    response = frappe.new_doc("Questionnaire Response")

    response.questionnaire = questionnaire
    response.branch = branch
    response.auditor_name = auditor_name
    response.audit_month = audit_month
    response.audit_dates = audit_dates
    response.unit_audited = unit_audited
    response.gstin_of_unit = gstin_of_unit
    response.nature_of_operations = nature_of_operations
    response.persons_met_during_audit = persons_met_during_audit
    response.previous_audit_pending_points = previous_audit_pending_points
    response.serious_irregularities_found = serious_irregularities_found
    response.fraud_suspected = fraud_suspected
    response.matter_requiring_immediate_attention = matter_requiring_immediate_attention
    response.submitted_by = frappe.session.user
    response.employee_id = employee_id
    response.employee_name = employee_name
    response.department = department

    for row in answers:
        if row.get("field_type") == "Table":

            for idx, table_row in enumerate(row.get("table", []), start=1):

                response.append("questionnaire_table_response",{
                    "question": row["question"],
                    "row_no": idx,
                    "row_data": json.dumps(table_row)
                })

        else:
            response.append("questionnaire_answer", {
                "question": row["question"],
                "answer": row["answer"],
                "rating": row.get("rating") or 0,
                "remark": row.get("remark")
            })

        # frappe.throw(f"answers: {row}") 
    response.insert(ignore_permissions=True)

    frappe.db.commit()

    return response.name

@frappe.whitelist(allow_guest=True)
def get_child_table_fields(doctype):

    meta = frappe.get_meta(doctype)

    fields = []

    for df in meta.fields:

        if df.fieldtype in (
            "Data",
            "Int",
            "Float",
            "Currency",
            "Date",
            "Select",
            "Check",
            "Link",
            "Small Text"
        ):

            fields.append({
                "label": df.label,
                "fieldname": df.fieldname,
                "fieldtype": df.fieldtype,
                "options": df.options
            })

    return fields


@frappe.whitelist(allow_guest=True)
def get_questionnaires():
    return frappe.get_all("Questionnaire Template", fields=["name"], order_by="name" )


@frappe.whitelist(allow_guest=True)
def get_responses(questionnaire):
    return frappe.get_all("Questionnaire Response",
        filters={
            "questionnaire": questionnaire
        },
        fields=[
            "name",
            "auditor_name",
            "branch",
            "audit_dates",
            "audit_month",
            "employee_name"
        ],
        order_by="creation desc"
    )

import frappe
import json

@frappe.whitelist(allow_guest=True)
def get_response(response):

    response_doc = frappe.get_doc("Questionnaire Response", response)

    template = frappe.get_doc(
        "Questionnaire Template",
        response_doc.questionnaire
    )

    sections = frappe.get_all(
        "Questionnaire Section",
        filters={
            "questionnaire": template.name
        },
        fields=[
            "name",
            "section_name",
            "sequence",
            "tab",
            "depends_on",
            "sub_section"
        ],
        order_by="sequence"
    )

    questions = frappe.get_all(
        "Questionnaire Question",
        filters={
            "questionnaire": template.name
        },
        fields=[
            "*"
        ],
        order_by="sequence"
    )

    answers = {}

    for d in response_doc.questionnaire_answer:
        answers[d.question] = {
            "answer": d.answer,
            "remark": d.remark,
            "rating": d.rating
        }

    tables = {}

    for d in response_doc.questionnaire_table_response:

        if d.question not in tables:
            tables[d.question] = []

        tables[d.question].append(json.loads(d.row_data))

    return {

        "response": response_doc,
        "sections": sections,
        "questions": questions,
        "answers": answers,
        "tables": tables

    }